pkg/capabilities/oci/manifest/manifest.go

```diff
@@ -0,0 +1,46 @@
+package manifest
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/kubewarden/policy-sdk-go/constants"
+	cap "github.com/kubewarden/policy-sdk-go/pkg/capabilities"
+	specs "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// GetOCIManifest fetches the OCI manifest for the given image URI.
+// The returned value, depends of the given image. It could be a OCI image manifest
+// or a OCI index image manifest. See more at:
+// https://github.com/opencontainers/image-spec/blob/main/manifest.md
+// https://github.com/opencontainers/image-spec/blob/main/image-index.md
+// Arguments:
+// * image: image to be verified (e.g.: `registry.testing.lan/busybox:1.0.0`)
+func GetOCIManifest(h *cap.Host, image string) (OciImageManifestResponse, error) {
+	response := OciImageManifestResponse{}
+	// build request payload, e.g: `"ghcr.io/kubewarden/policies/pod-privileged:v0.1.10"`
+	payload, err := json.Marshal(image)
+	if err != nil {
+		return response, fmt.Errorf("cannot serialize image URI to JSON: %w", err)
+	}
+
+	// perform host callback
+	responsePayload, err := h.Client.HostCall("kubewarden", "oci", "v1/oci_manifest", payload)
+	if err != nil {
+		return response, err
+	}
+
+	imageManifest := specs.Manifest{}
+	indexManifest := specs.Index{}
+	imageErr := json.Unmarshal(responsePayload, &imageManifest)
+	indexErr := json.Unmarshal(responsePayload, &indexManifest)
+	if imageErr == nil && imageManifest.MediaType == specs.MediaTypeImageManifest || imageManifest.MediaType == constants.ImageManifestMediaType {
+		response.image = &imageManifest
+	} else if indexErr == nil && imageManifest.MediaType == specs.MediaTypeImageIndex || indexManifest.MediaType == constants.ImageManifestListMediaType {
+		response.index = &indexManifest
+	} else {
+		return response, fmt.Errorf("response media type not supported: %s", imageManifest.MediaType)
+	}
```

Comment:
This is not handling the case where neither `imageManifest` nor `indexManifest` cannot be unmarshalled. Also, the `else` statements assumes the `MediatType` has to be taken from `imageManifest`; however this is not always true. Consider the case where the response can be converted into a `image.Index` and not into a `image.Manifest`, plus the `image.Index.MediaType` doesn't match the allowed values. In this case `indexManifest` would be `nil` and the code would crash.

This is one of the traps of doing all these if/else if/else branches. I propose this approach. First of all remove the declaration of `response` right at the top of the method. Then write this highlighted section of the code like that:


```golang
imageManifest := specs.Manifest{}
if err := json.Unmarshal(responsePayload, &imageManifest); err == nil {
   // this is a new func to be written
   if isValidMediaType(imageManifest.MediaType) {
     response := OciImageManifestResponse{
        image: &imageManifest
     }
     return response, nil
   }
   // yes, change the signature of the function to return  (ptr, err)
   return nil, errors.New("not a valid media type")
}

indexManifest := specs.Index{}
if err := json.Unmarshal(responsePayload, &indexManifest); err == nil {
   if isValidMediaType(indexManifest.MediaType) {
     response := OciImageManifestResponse{
        index: &indexManifest
     }
     return response, nil
   }
   return nil, errors.New("not a valid media type")
}

return nil, errors.New("cannot decode response")
```



```

Reply:
This will not work due how the go JSON unmarshaling works. It will always consider the index manifest as an image manifest. In other words, ignoring the bad formatted JSON cases, the unmarshaling for the image  and index manifest always succeed. Despite of  the payload.  That's why it's checking only the image manifest media type. Both struct has this field, therefore, does not matter if I check the first (image manifest). If I applied your changes, the index payload will never be parsed. Take a look:

```go
func isValidMediaType(mediaType string) bool {
	return mediaType == specs.MediaTypeImageManifest ||
		mediaType == constants.ImageManifestMediaType ||
		mediaType == specs.MediaTypeImageIndex ||
		mediaType == constants.ImageManifestListMediaType
}

// GetOCIManifest fetches the OCI manifest for the given image URI.
// The returned value, depends of the given image. It could be a OCI image manifest
// or a OCI index image manifest. See more at:
// https://github.com/opencontainers/image-spec/blob/main/manifest.md
// https://github.com/opencontainers/image-spec/blob/main/image-index.md
// Arguments:
// * image: image to be verified (e.g.: `registry.testing.lan/busybox:1.0.0`)
func GetOCIManifest(h *cap.Host, image string) (*OciImageManifestResponse, error) {
	// build request payload, e.g: `"ghcr.io/kubewarden/policies/pod-privileged:v0.1.10"`
	payload, err := json.Marshal(image)
	if err != nil {
		return nil, fmt.Errorf("cannot serialize image URI to JSON: %w", err)
	}

	// perform host callback
	responsePayload, err := h.Client.HostCall("kubewarden", "oci", "v1/oci_manifest", payload)
	if err != nil {
		return nil, err
	}

	imageManifest := specs.Manifest{}
	if err := json.Unmarshal(responsePayload, &imageManifest); err == nil {
		// this is a new func to be written
		if isValidMediaType(imageManifest.MediaType){
			response := OciImageManifestResponse{
				image: &imageManifest,
			}
			return &response, nil
		}
		return nil, fmt.Errorf("not a valid media type: %s", imageManifest.MediaType)
	}

	indexManifest := specs.Index{}
	if err := json.Unmarshal(responsePayload, &indexManifest); err == nil {
		if isValidMediaType(indexManifest.MediaType) {
			response := OciImageManifestResponse{
				index: &indexManifest,
			}
			return &response, nil
		}
		return nil, fmt.Errorf("not a valid media type: %s", indexManifest.MediaType)
	}

	return nil, errors.New("cannot decode response")

}
```

The tests results:

```
go test ./... -coverprofile cover.out
?       github.com/kubewarden/policy-sdk-go/constants   [no test files]
?       github.com/kubewarden/policy-sdk-go/mock/capabilities   [no test files]
?       github.com/kubewarden/policy-sdk-go/pkg/capabilities    [no test files]
?       github.com/kubewarden/policy-sdk-go/pkg/capabilities/oci        [no test files]
ok      github.com/kubewarden/policy-sdk-go     0.003s  coverage: 42.4% of statements
ok      github.com/kubewarden/policy-sdk-go/pkg/capabilities/crypto     0.001s  coverage: 66.7% of statements
ok      github.com/kubewarden/policy-sdk-go/pkg/capabilities/kubernetes 0.001s  coverage: 71.4% of statements
ok      github.com/kubewarden/policy-sdk-go/pkg/capabilities/net        0.002s  coverage: 70.0% of statements
?       github.com/kubewarden/policy-sdk-go/protocol    [no test files]
?       github.com/kubewarden/policy-sdk-go/testing     [no test files]
--- FAIL: TestV1OciIndexManifest (0.00s)
    --- FAIL: TestV1OciIndexManifest/application/vnd.oci.image.index.v1+json (0.00s)
        manifest_test.go:101: response should not be an image manifest:
            &{{0} application/vnd.oci.image.index.v1+json {  0 [] map[] <nil>} [] map[annonation:annotationValue]}
    --- FAIL: TestV1OciIndexManifest/application/vnd.docker.distribution.manifest.list.v2+json (0.00s)
        manifest_test.go:101: response should not be an image manifest:
            &{{0} application/vnd.docker.distribution.manifest.list.v2+json {  0 [] map[] <nil>} [] map[annonation:annotationValue]}
FAIL
coverage: 59.1% of statements
FAIL    github.com/kubewarden/policy-sdk-go/pkg/capabilities/oci/manifest       0.002s
ok      github.com/kubewarden/policy-sdk-go/pkg/capabilities/oci/manifest_digest        0.001s  coverage: 70.0% of statements
ok      github.com/kubewarden/policy-sdk-go/pkg/capabilities/oci/verify_v1      0.001s  coverage: 100.0% of statements
ok      github.com/kubewarden/policy-sdk-go/pkg/capabilities/oci/verify_v2      0.001s  coverage: 100.0% of statements
FAIL
```

Furthermore, if I replace the `isValidMediaType` function to a `if` where it check only the allowed media type considering the manifest type unmarshaled, that also failed. Because go json unmarshal will always be able to unmarshal the index into the image manifest.

Reply:
@flavio, I've pushed a code inspired in out code change. Please, take a look on it again.